src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -427,10 +427,9 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
             MethodSymbol cacheKey = GetCacheKey(delegateType);
 
             BoundLambda returnInferenceLambda;
-            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
+            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType && returnInferenceLambda.Symbol.ReturnType == returnType)
             {
                 lambdaSymbol = returnInferenceLambda.Symbol;
-                Debug.Assert(lambdaSymbol.ReturnType == returnType);
                 Debug.Assert(lambdaSymbol.RefKind == refKind);
```

Comment:
Actually, since a ref returning lambda requires an explicit `ref` keyword, it may not be possible for a lambda to be bound successfully with different `RefKind`. In short, we can leave the `RefKind` check as an assert.